First pass on conformance validation

mayankshah1607 opened the pull request:
This PR adds the following changes into the repository:

- Add a single test for verifying proxy injection
- Adds test helpers for installing / uninstalling control plane and running pre and post installation checks
- Adds support for reading test configuration from a config YAML
- Add a Dockerfile for housing the conformance tests. For the sake of testing, this image is accessible on [hub.docker.com](hub.docker.com) at [mayankshah1607/linkerd2-conformance](https://hub.docker.com/repository/docker/mayankshah1607/linkerd2-conformance)
- Add convenience scripts for building and pushing the various docker images for local development, as well as for running the tool locally
-  Add a Sonobuoy plugin file
- Update README with relevant documention regarding description, use cases and usage

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update dependencies

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add `conformance_test.go`

- The `conformance_test.go` file acts as an entrypoint into the test
suite - a runner for the specs by ginkgo

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add specs for `linkerd install`

- Added Ginkgo defined specs for linkerd install, that runs a single
test to see if `linkerd install` run successfully

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add helpers

- Add helpers that reuse the `linkerd/testutil` package - in particular,
the TestHelpers for Linkerd integration tests, flags and Kubernetes
helpers

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add base image for conformance

- This base image provides `kubectl` for the purpose of reusability
along with the Sonobuoy image and the main image that houses the tests

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add Dockerfile

- This image holds our tests. This base image is intended to be used
either locally (by overriding the command) or along with Sonobuoy
plugin.

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add Sonobuoy plugin

- The `linkerd2-conformance.yaml` file defined the plugin that is
intended to be used with the Sonobuoy CLI.

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add image for Sonobuoy

- The `sonobuoy.Dockerfile` is intended to be used for the sake of
running Sonobuoy from within a Docker container. This allows users to
run the Sonobuoy plugin without depending on installing the Sonobuoy CLI
locally.

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Refactor Sonobuoy items

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update README

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add helper scripts

- These helper scripts are mainly responsible for running commands for
building and pushing the various docker images
- Additional helper script for running `go test`

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add LICENSE

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: Fix typo in README

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 commented:
Ping @Pothulapati

Pothulapati commented on README.md:
```diff
@@ -1,2 +1,101 @@
+$ curl -LJO https://raw.githubusercontent.com/mayankshah1607/linkerd2-conformance/master/sonobuoy/plugin.yaml
+
+# Run the plugin
+$ sonobuoy run --plugin plugin.yaml --wait
```
Looks like `--plugin` can take a URL directly.

mayankshah1607 replied:
Yep! Downloading the plugin gives users the freedom to set their own options in the plugin definition. For example, the start up command as defined in the plugin currently uses a set of flags. Users could append flags like `--ginkgo.progress` to show progress reports on the pod logs, or `-linkerd-namespace` to modify the control plane ns before kicking off sonobuoy.

But for folks who want to use it directly out of the box, we should document that the plugin URL can be used directly! :)

mayankshah1607 commented on README.md:
```diff
@@ -1,2 +1,101 @@
+
+These commands assume that the user has the [Sonobuoy CLI](https://github.com/vmware-tanzu/sonobuoy#installation) and [kubectl](https://kubernetes.io/docs/reference/kubectl/overview/) (with a correctly configured `kubeconfig`) installed locally.
+
+This repo provides a Sonobuoy plugin file that is intended to be plugged into the Sonobuoy CLI. Sonobuoy reads the plugin definition, and spins up a pod with the [linkerd2-conformance Docker image]().
```
The link for this image is in the PR description. Left it empty here for now so that we can have the link from Linkerd2's container registry.

Pothulapati commented on specs/install/specs.go:
```diff
@@ -0,0 +1,59 @@
+			out, err := h.KubectlApply(out, "")
+
+			gomega.Expect(err).To(gomega.BeNil())
+
```
Why not perform `linkerd check` and report its output? Would complete a total Linkerd install check and should be a pretty small change.

mayankshah1607 replied:
Yep! That was actually the next test I was going to write - to perform `check`, and also wait for the control plane components to change state to `Running`.

Will add that now :)

Pothulapati reviewed (commented):
Awesome first pass! :100: 

I have the following suggestions/thoughts

- Users would expect the test to leave the cluster in the same state as to when it was created. I see that we don't clean up the installed resources in this pass yet. Should we have a global clean that we run before moving the results into the volume? or creates tests such that each test cleans up what it has created.
- I also think just like users would be able to try Linkerd at different k8s cluster configurations, This would also help us to test conformance with different Linkerd binaries i.e edges, development ones, etc to conform if everything is correct. It would be really cool to have a easier UX to pass the binary into sonobuoy. Allowing Users to pass the binary version seems like one good answer. If a custom binary has to be sent I guess the option would be to create it as a volume and update the plugin file, which isn't that bad. WDYT? Can we do anything better here? This can be part of the config file that we discussed.
- I see that ginkgo can export to `JUnit` and the same can be further used by sonobuoy to render really clean outputs without having to extract tar files and stuff. 
Here I updated the `sonobuoy-config.result-format` to `junit` (replacing raw which was default) and sonobuoy was able to render the following

```yaml
 sonobuoy results ./linkerd2-conformance/202006011213_sonobuoy_2841152d-ee08-4d67-8e89-5029c0ae5edf.tar.gz --plugin linkerd2-conformance --mode detailed | jq
{
  "name": "Verifying linkerd2 control plane components By attempting to installing control plane components should successfully pipe output of `linkerd install` to `kubectl`",
  "status": "passed",
  "meta": {
    "path": "linkerd2-conformance|report.out.xml|control plane installation spec"
  }
}
``` 
The above output seems good, unless we have some problem going this way! If we are planning to with the above output, we should figure out a standard that we would use for each test i.e where would checks descriptions go and stuff and would be nice to have them be common across all tests. Following the [ginkgo's docs](https://onsi.github.io/ginkgo/#structuring-your-specs) in structuring the descriptions/tests seems to be the answer here to have consistent outputs!

mayankshah1607 commented:
@Pothulapati Thank you! Those suggestions are really valuable.

- The `linkerd2/testutil` does provide an `uninstall` flag, so we can use its value to check if an uninstall must be performed, which was going to be a part of the next pass. However, I feel a global uninstall should be optional - if any of the tests fail, the users may want to actually dive into this failed state of the cluster and figure certain things out - such as viewing logs from the control plane or the proxy, or verifying the state of the workloads.  This could be the case for a successful pass on the tests too. WDYT?

- Yep, passing a custom version of the binary makes a lot of sense. As of this pass, we're pretty tightly coupled with the base image provided (which always fetches the latest stable release). Let me think this through and see what options we have available, and get back to you :)

- This form of an output seems great! For now, I think we should definitely go with this, but later on, we may have to consider how we'll let the users pick the format they want. While having them to edit the plugin file manually seems fine, we could have a CLI command that accepts these options, and renders a plugin.yaml that can be plugged into Sonobuoy. WDYT?

mayankshah1607 committed:
Update Dockerfile

- Remove redundant base.Dockerfile and consolidate into Dockerfile
- Remove sonobuoy.Dockerfile (idea needs rework)
- Update Dockerfile to not install linkerd2 binary

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add test for verifying state of control plane

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add spec for uninstall

- Refactor specs

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add check tests

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update scripts

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update Sonobuoy workflow

- Move sonobuoy command to separate `run.sh`
- Update Dockerfile

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Clear directory

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update instructions on Sonobuoy

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 commented:
@Pothulapati As per the comments, I've updated this PR for the first pass:

- Along with the install tests, added tests for uninstall, and `linkerd check`
- Decouple the linkerd2 binary installation process from the Dockerfile. Now the users have an option to choose the version of the linkerd2 binary to test against. 
- Update instructions on running with Sonobuoy to have a nice JSON output without much hassle.  

I haven't yet added instructions on how to run the tests inside a stand-alone container as I'm expecting a change in the flow after https://github.com/linkerd/linkerd2/pull/4532 gets merged

mayankshah1607 commented on README.md:
```diff
@@ -1,2 +1,102 @@
+
+$ sonobuoy run \
+  --plugin $plugin \
+  --plugin-env linkerd2-conformance.LINKERD2_VERSION=[VERSION]\
```
@Pothulapati Looks like Sonobuoy lets you pass environment variables for the pod during runtime. Fortunately, the [script](https://raw.githubusercontent.com/linkerd/website/master/run.linkerd.io/public/install) used for installing the linkerd2 binary uses the `LINKERD2_VERSION` environment variable to fetch the required binary of a particular version.
This seems like a neat option to me. WDYT? :)

mayankshah1607 commented on run.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+# This script is to be run from inside the Sonobuoy pod
+
+curl -sL https://raw.githubusercontent.com/linkerd/website/master/run.linkerd.io/public/install | sh
+
+/bin/bash -c "./conformance -integration-tests -linkerd /root/.linkerd2/bin/linkerd -ginkgo.v -linkerd-namespace l5d-conformance --ginkgo.reportFile=/tmp/results/report.xml"
+
+
+tar czvf /tmp/results/results.tar.gz -C /tmp/results .
+
+echo -n /tmp/results/results.tar.gz > /tmp/results/done
```
I've introduced this script after going through the Sonobuoy docs. Instead of cluttering the plugin file with a lot of commands, the container simply uses this as a start up script to install the specified version of linkerd2 binary (which is read from the `LINKERD2_VERSION` env variable), and initiate the tests.

Pothulapati commented on README.md:
```diff
@@ -1,2 +1,102 @@
+
+This repo contains the conformance tests for [Linkerd2](https://github.com/linkerd/linkerd2) as described by [this](https://github.com/linkerd/rfc/pull/24) RFC.
+
+The conformance validation tool is primarily intended to be run against an installation of Linkerd2 to verify the correctness of a [Kubernetes](https://kubernetes.io/) cluster's configuration with respect to Linkerd as well as validate non-trivial network communication (HTTP, gRPC, websocket) among stateless and stateful workloads in the Linkerd data plane.
```
Hmm, this gives an impression that you would need Linkerd install and would run conformance on top but we install Linkerd too right. So, How about "run on a specific Linkerd binary"?

Pothulapati commented on plugin.yaml:
```diff
@@ -0,0 +1,28 @@
+  command:
+  - /bin/bash
+  args:
+      - ./run.sh
```
Hmm, Do we need `run.sh` actually? By removing it we can directly pass the planned config yaml here which is confused by a conformance helper and downloads the binary, etc and then calls the `testutil.TestHelper` with relevant configuration.

By removing this we only have a single layer i.e Go Conformance Helpers where we perform book-keeping tasks initially.. I think we can remove it unless there is some use-case. :)

mayankshah1607 replied:
It exists only to make sure that the command section is not too untidy and filled with commands. Once we get https://github.com/linkerd/linkerd2/pull/4532 merged, I'll make sure to remove the binary installation process from here and make it a part of the config yaml. :smile:

Pothulapati reviewed (commented):
I could be doing something wrong but when i run ` sonobuoy run --plugin ./plugin.yaml --plugin-env linkerd2-conformance.LINKERD2_VERSION=stable-2.8`. The status is just marked as completed and reading the pod logs I see
```
Downloading linkerd2-cli-stable-2.8-linux...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
curl: (22) The requested URL returned error: 404 Not Found
-linkerd binary does not exist
./
tar: .: file changed as we read it
```

A question I have, Also what defines a test `spec`? Atleast in the Linkerd repo, we each test life-cycle into a single atomic method (not atomic atomic, but atleast in the code structuring)? Here we have each function into a spec? Does that provide better structure in outputs? If it does, maybe it could also help us in re-using specs like check across multiple tests?

mayankshah1607 commented:
@Pothulapati each "spec" corresponds to one feature of Linkerd2. Each of these specs further have multiple test cases, where each of these test cases are wrapped inside a single `It` block. I thought it would be nice to have a separate file which defines all the test cases, each as a single function, and call them inside the `It` block just so that the `Context` block is not cluttered with a lot of code. This was done just to make sure that the code is readable and organized well.
For example, each of the featues have a `spec.go` which acts as the entry point of the test, and the main body of the various test cases is defined in `tests.go`.

Hence, the `spec.go` is where the single feature is structured into multiple test cases, and the `tests.go` is where each of the tests are defined and where assertions happens. :smile:

mayankshah1607 committed:
feat: read test options from config yaml

- Add new ConformanceTestOptions to store conformance test options
- Initialize helper using `testutil.NewGenericTestHelper`
- Add helpers to parse defautls

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update check tests

- Parse JSON output obtained from check, to scan for errors

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Minor refactoring

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Refactor to use configMap

- Update plugin.yaml to use the mounted config.yaml
- refactor sonobuoy items

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add helper to install linkerd

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Refactoring specs

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Remove config.yaml

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add sample config

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update docs for Sonobuoy

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: fix typo

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: fix minor nits

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Refactor test structure

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Bump linkerd/linkerd2 version

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Refactor code for using new `NewGenericTestHelper` method

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
check: update test to use validate expected version

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Refactor constants

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
config: fetch and use latest stable when version is unspecified

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
refactor: use single line error handling

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
config: make configMap optional

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Remove .tar file added by accident

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Bump ginkgo version

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Remove plugin.txt

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update GoDoc

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Bump linkerd2 version

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Refactor test structure and add inject tests

- `spec/` now holds primary specs as defined by the RFC
- move install, uninstall, check tests as helper gomega assertions for
reusing in primary tests
- add inject specific configuration in ConformanceTestOptions
- add basic inject tests
- refactor logic to parse defaults

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Normalize By messages

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Improve By messages

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update run.sh to capture pod logs

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update Dockerfile

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update Dockerfile

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update docs

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update docs

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add tests for proxy container injection

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Improve failure annotations in test_helpers.go

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add inject test to clean resources created during testing

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Remove results

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add separate specs for install and uninstall

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update Go imports

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Fixed a bug in globalControlPlane install logic

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

olix0r commented on config.yaml:
```diff
@@ -0,0 +1,18 @@
+                    request: 10m
+                memory:
+                    limit: 100Mi
+                    request: 50Mi:W
```
typo ;)

mayankshah1607 replied:
Oops! If only I could be a little more careful while using vim..

mayankshah1607 committed:
Fix typo

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add HA option for install

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
refactor: make code easier to read

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update inject test to test for ns level overrides

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Fix go.mod

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Refactor variable names and fix typo

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Decouple install logic from TestHelper init

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Improve error handling in parse()

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Improve code readability and consistency

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Refactor binary install logic

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add upgrade specs

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
feat: users can now specify install flags in config.yaml

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: Add repository structure

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: Add test configuration docs

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: add backticks for flags

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: add section on how to add new tests

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: update config table

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: add table of contents

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: move repository structure

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

Pothulapati commented on utils/config.go:
```diff
@@ -0,0 +1,234 @@
+var (
+	defaultNs            = "l5d-conformance"
+	defaultClusterDomain = "cluster.local"
+	defaultVersion       = "stable-2.8.0"
```
I don't see a reason to set the default version. We can get the default version of Linkerd2 when we we don't set the env right? Can we do something with that here?

Pothulapati replied:
Also, How about using the latest edge as the default?

mayankshah1607 replied:
We're actually doing that only - the latest version of Linkerd is being fetched from the `version.json` endpoint if you have a look at the `parse()` method; however it's possible for that to return an error for any reason, and hence it would just use this statically set `defaultVersion` :)
It still makes sense to remove this variable because in that case, user should just manually set a version

Pothulapati replied:
Yeah. I don't see why fetch version.json could fail as we heavily depend on it in a lot of other places. We can just remove this and propagate a error! if both are not available i.e version.json and users override.

Pothulapati commented on config.yaml:
```diff
@@ -0,0 +1,28 @@
+                    memory:
+                        limit: 100Mi
+                        request: 50Mi
+    globalControlPlane:
```
I don't see why `globalControlPlane` would come here as we were thinking that flag would be more like a global one which acts like the default install tests.

Something like
```yaml
linkerdVersion: stable-2.8.1
externalIssuer: false
globalControlPlane:
    enable: true
    uninstall: true
install:
    enabled: true
    config:
        ha: false
        # upgradeFromVersion: stable-2.7.0
        flags:
            - "--controller-log-level"
            - "debug"
            - "--proxy-log-level"
            - "warn,linkerd2_proxy=debug"
        addOns:
            tracing:
                enabled: true
                collector:
                    resources:
                        cpu:
                            limit: 100m
                            request: 10m
                        memory:
                            limit: 100Mi
                            request: 50Mi
inject:
    enabled: true
    clean: true
    config:
         flags:
...

```

Also, What do you think about each test having `skipTest` and having some common fields across all tests and extra config field based on the test?

mayankshah1607 replied:
Hmm, the reason for choosing `skipTest` was that the zero value of booleans is `false`, which means, if the user chooses to skip the `enabled` field, it will always default to `false` and the test will not run unless they always manually specify `enabled: true`

Pothulapati reviewed (commented):
I tried to run a clean test, and my tests failed twice with the following errors

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   644  100   644    0     0   1638      0 --:--:-- --:--:-- --:--:--  1634

  0 37.0M    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
  0 37.0M    0  134k    0     0  57354      0  0:11:17  0:00:02  0:11:15  145k
  4 37.0M    4 1631k    0     0   490k      0  0:01:17  0:00:03  0:01:14  885k
 34 37.0M   34 12.9M    0     0  2962k      0  0:00:12  0:00:04  0:00:08 4426k
 66 37.0M   66 24.6M    0     0  4682k      0  0:00:08  0:00:05  0:00:03 6452k
 96 37.0M   96 35.7M    0     0  5793k      0  0:00:06  0:00:06 --:--:-- 7566k
100 37.0M  100 37.0M    0     0  5996k      0  0:00:06  0:00:06 --:--:-- 9631k
panic: 
Your test failed.
Ginkgo panics to prevent subsequent assertions from running.
Normally Ginkgo rescues this panic so you shouldn't see it.

But, if you make an assertion in a goroutine, Ginkgo can't capture the panic.
To circumvent this, you should call

	defer GinkgoRecover()

at the top of the goroutine that caused this panic.
 [recovered]
	panic: 
Your test failed.
Ginkgo panics to prevent subsequent assertions from running.
Normally Ginkgo rescues this panic so you shouldn't see it.

But, if you make an assertion in a goroutine, Ginkgo can't capture the panic.
To circumvent this, you should call

	defer GinkgoRecover()

at the top of the goroutine that caused this panic.


goroutine 67 [running]:
testing.tRunner.func1.1(0x1534de0, 0x1977040)
	/usr/local/go/src/testing/testing.go:940 +0x2f5
testing.tRunner.func1(0xc0004410e0)
	/usr/local/go/src/testing/testing.go:943 +0x3f9
panic(0x1534de0, 0x1977040)
	/usr/local/go/src/runtime/panic.go:969 +0x166
github.com/onsi/ginkgo.Skip(0x178cf67, 0x16, 0x0, 0x0, 0x0)
	/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/ginkgo_dsl.go:249 +0xc8
github.com/linkerd/linkerd2-conformance/specs/upgrade.RunUpgradeSpec.func1()
	/conformance/specs/upgrade/spec.go:14 +0xeb
github.com/onsi/ginkgo/internal/suite.(*Suite).PushContainerNode(0xc0000d0f00, 0x1786ccf, 0x11, 0x182bfc8, 0x0, 0x24d5905, 0x22, 0xa, 0xc000899100, 0x635)
	/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/internal/suite/suite.go:147 +0x158
github.com/onsi/ginkgo.Describe(0x1786ccf, 0x11, 0x182bfc8, 0x1)
	/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/ginkgo_dsl.go:287 +0xaa
github.com/linkerd/linkerd2-conformance/specs/upgrade.RunUpgradeSpec(0x1)
	/conformance/specs/upgrade/spec.go:10 +0x42
github.com/linkerd/linkerd2-conformance/specs.runPreFlightSpecs.func1()
	/conformance/specs/specs.go:26 +0x39
github.com/onsi/ginkgo/internal/suite.(*Suite).PushContainerNode(0xc0000d0f00, 0x0, 0x0, 0xc0003a9da0, 0x0, 0x24d58c6, 0x1b, 0x15, 0xc00068d400, 0x3e8)
	/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/internal/suite/suite.go:147 +0x158
github.com/onsi/ginkgo.Describe(0x0, 0x0, 0xc0003a9da0, 0x0)
	/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/ginkgo_dsl.go:287 +0xaa
github.com/linkerd/linkerd2-conformance/specs.runPreFlightSpecs(0xc0002c0b00, 0xc0001a9680, 0x1)
	/conformance/specs/specs.go:21 +0x4e
github.com/linkerd/linkerd2-conformance/specs.RunAllSpecs.func1()
	/conformance/specs/specs.go:66 +0x3d
github.com/onsi/ginkgo/internal/suite.(*Suite).PushContainerNode(0xc0000d0f00, 0x0, 0x0, 0xc0003a9f40, 0x0, 0x24d58c6, 0x1b, 0x41, 0xc0002081a0, 0x18a)
	/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/internal/suite/suite.go:147 +0x158
github.com/onsi/ginkgo.Describe(0x0, 0x0, 0xc0003a9f40, 0x39094b65)
	/go/pkg/mod/github.com/onsi/ginkgo@v1.13.0/ginkgo_dsl.go:287 +0xaa
github.com/linkerd/linkerd2-conformance/specs.RunAllSpecs(0xc0004410e0)
	/conformance/specs/specs.go:65 +0x69
github.com/linkerd/linkerd2-conformance_test.TestConformance(0xc0004410e0)
	/conformance/conformance_test.go:39 +0x2b
testing.tRunner(0xc0004410e0, 0x182bfe0)
	/usr/local/go/src/testing/testing.go:991 +0xdc
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1042 +0x357
./
./summary
tar: .: file changed as we read it
```

and 

the other time I got the same `file changed as we read` but at a different point in execution.

I did not do anything different. Just followed the ReadMe. No Linkerd installation happened in both the cases.

mayankshah1607 committed:
Fix bug due to `Skip`

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Remove accidentally added addons file

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

Pothulapati commented:
It works now! :clap:

mayankshah1607 committed:
Bump linkerd2 version

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update parameters for NewGenericTestHelper

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
new format for config.yaml

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: update config fields

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
major refactor

- Refactor tests to have a separate test suite (`_test.go` file), rather
than a spec under a single suite
- New Dockerfile: remove multi stage build as its no longer nesessary
- New scripts for building/pushing image, running single and multiple
tests
- Refactor install/uninstall/upgrade tests into single `lifecycle`

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Remove results file

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Fix bug in script

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Fix Dockerfile

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update plugin

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: Update info on local development

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: Update info on how to add new tests

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: fix typo

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update README.md

- Update "Table of Contents"
- Update "Repository structure"
- Update commands for running locally
- Fix minor typos

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: fix typo

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update defaults while parsing

- remove `defaultVersion` variable
- error out when failing to fetch latest edge
- Use `defaultPath`

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add ingress-nginx test

- Add a new test suite for ingress: only tests using the nginx
controller for now
- Add test utils for emojivoto
- refactor consts into a single file

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add ingress test to `run-all`

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update run script to continue if a test fails

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
ingress: update tests to read config from YAML

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
inject: make proxy injection process retryable

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

ihcsim commented on config.yaml:
```diff
@@ -0,0 +1,35 @@
+      #                 memory:
+      #                     limit: 100Mi
+      #                     request: 50Mi
+lifecycle:
+    skip: false
+    reinstall: false 
+    uninstall: true
+    # upgradeFromVersion: stable-2.7.0
+inject:
+    skip: false
+    clean: true
+ingress:
+    skip: false
+    config:
+        controllers:
+            - nginx
```
I think we should nest these test case configuration, under a parent variable named `testCase`. This will separate the test case configuration from the general meta configuration like Linkerd version, binary path,

E.g.,

```yaml
testCase:
  lifecycle:
    skip: false
    reinstall: false 
    uninstall: true   
  inject:
    skip: false
    clean: true
  ingress:
    skip: false
    config:
      controllers:
       - nginx
```

ihcsim commented on config.yaml:
```diff
@@ -0,0 +1,35 @@
+      #                     limit: 100Mi
+      #                     request: 50Mi
+lifecycle:
+    skip: false
```
Instead of specifying `skip: [false|true]` for each test case, what do you think of utilizing regex to pattern match test functions to be skipped?

E.g.,

```yaml
testSuite:
   skip: ["TestIngress.*", "^MyFlakyTest$"]
```

mayankshah1607 replied:
Utiizing regex patterns would be useful if these could be passed down to the `--skip` flag of the [Ginkgo CLI](https://onsi.github.io/ginkgo/#the-ginkgo-cli). But in our case, we're not programatically determining the values for any of the CLI flags, and neither does Ginkgo natively provide APIs to skip tests on the basis of regex patterns.
To programatically determine if a Ginkgo test should be skipped would mean having to check a condition under the `ginkgo.BeforeEach` block and calling `ginkgo.Skip`. I don't see any advantage over matching regex patterns in the code than simply determining boolean values to evaluate an `if` condition. Moreover, we also have to consider that our users will not be familiar with the name of the test functions in our code (like `TestIngress`, etc) when they choose to run our Sonobuoy plugin - they only have to deal with setting up a correct config YAML and running the plugin. We could document these names, but I'm not really sure if this makes our user experience better.

Ginkgo also works little differently from the standard `testing` library. Each of our test suites, actually has only one `Test*` function, which holds several [Ginkgo 'containers'](https://onsi.github.io/ginkgo/#structuring-your-specs). For example, the ingress test suite contains only one function `TestIngress` that holds our Ginkgo specs, and similarly for inject (TestInject), etc. Using regex for such similar names seems a little superfluous to me. 

If the redundancy of the `skip` flag is an issue, what do you think about having to skip on the basis of test names rather than regex pattern? For example:
```yaml
testSuite:
  skip:
    - ingress
    - inject
    - myFlakyTest
```

ihcsim replied:
Can we just pass the regex to the `go test` (adding a `-run` flag) in the `bin/go-test`? (I assume that's the entrypoint to run the test.) Feels like a lot of boiler plate code to have to do this for every spec.

```
	ginkgo.BeforeEach(func() {
		if c.Skip<TestName>() {
			ginkgo.Skip("Skipping <test> tests")
		}
	})
```

mayankshah1607 replied:
Hmm, but this might not work too well when running with the Sonobuoy workflow where the scripts are hidden away from the users and all they have to do is define the configuration in the YAML and mount it as a configMap.
The config YAML is supposed to be the single source for reading any kind of configuration to have a consistent UX. What do you think?

mayankshah1607 replied:
For now, I've made this boilerplate a single reusable function :
```go
func ShouldTestSkip(skip bool, message string) bool {
	return ginkgo.BeforeEach(func() {
		if skip {
			ginkgo.Skip(message)
		}
	})
}
```

This helper is called in the beginning of every skippable test:
```go
var _ = ginkgo.Describe("`linkerd inject`", func() {
	_, config := utils.GetHelperAndConfig()
	_ = utils.ShouldTestSkip(config.SkipInject(), "Skipping inject tests")
        
       // remaining specs go here
})

I hope this solves temporarily solves the problem of redundant boilerplate code. :smile:

ihcsim commented on config.yaml:
```diff
@@ -0,0 +1,35 @@
+controlPlane:
+    # namespace: l5d-conformance
+    config:   
+        ha: false
+        flags:
+            - "--controller-log-level"
+            - "debug"
+            - "--proxy-log-level"
+            - "warn,linkerd2_proxy=debug"
```
Do these config apply to all, most or specify test cases? I'd think something like `ha`, `flags` are only relevant to control plane installation tests. Will it make sense to group them under the `lifecycle` configuration?

mayankshah1607 replied:
These config apply to all installations of linkerd. Currently, we support to ways of control plane installation : 
1. a "universal" control plane - the control plane is installed as a part of the "lifecycle test", and the same control plane installation is used for the remaining tests.
2. a "local" control plane - the control plane is installed and uninstalled as a part of the "lifecycle" tests. Each suceeding test will then install a control plane, run the tests and assertions, and finally uninstall once the test is complete.

These configs are common to both our above cases. I'm completely okay with moving these under "lifecycle". @Pothulapati wdyt?

Pothulapati replied:
Hmm, Here we wanted to support both the cases where a test would and would not need a specific linkerd installation. If a specific test would need a different installation, the test-case can perform its own right in a different namespace, WDYT? :thinking:

ihcsim commented on README.md:
```diff
@@ -1,2 +1,254 @@
+| Option | Description | Default value |
+|-|-|-|
+| `linkerdVersion` | The linkerd2 binary version to use | Latest stable release |
+| `linkerdBinaryPath` | If specified, the tests use the binary installed in the directory. It is recommended that this is left unspecified while using Sonobuoy or if upgrade tests are enabled | `$HOME/.linkerd2/bin/linkerd` | 
```
The `linkerdBinaryPath` doesn't seem to work for me, when I run the tests locally:

```bash
$ cat config.yaml| grep linkerdBinaryPath                                            
linkerdBinaryPath: /home/isim/.linkerd2/bin/linkerd  # specify default location in config.yaml

$ bin/run-all
failed to initialize test helper or config: failed to parse YAML: yaml: unmarshal errors:
  line 2: field linkerdBinaryPath not found in type utils.ConformanceTestOptionsFAIL    github.com/linkerd/linkerd2-conformance/tests/lifecycle 0.014s
FAIL
failed to initialize test helper or config: failed to parse YAML: yaml: unmarshal errors:
  line 2: field linkerdBinaryPath not found in type utils.ConformanceTestOptionsFAIL    github.com/linkerd/linkerd2-conformance/tests/inject    0.014s
FAIL
```

mayankshah1607 replied:
This is fixed!

ihcsim commented on sonobuoy/run:
```diff
@@ -0,0 +1,10 @@
+# This script is meant to be run from inside the Sonobuoy pod
+bin/run-all
+
+cp -r ./reports /tmp/results
```
Consider adding the `reports` folder to `.gitignore`, so that git won't track that folder during multiple test runs.

ihcsim commented on bin/run-all:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
`#!/usr/bin/env sh` will be more flexible on different systems. See scripts in the Linkerd2 repo's `bin` folder.

ihcsim commented on bin/run-all:
```diff
@@ -0,0 +1,12 @@
+
+# Add tests here
+bin/go-test inject 02
+bin/go-test ingress 03
```
After this test failed on my local cluster, the test script went on to delete my Linkerd control plane. I am not sure how to debug why the tests failed. Any thoughts?

```
$ bin/run-all
....
=== RUN   TestIngress
Running Suite: ingress
======================
Random Seed: 1594265401
Will run 1 of 1 specs

STEP: Installing emojivoto
STEP: Injecting emojivoto
linkerd
  can work with nginx ingress controller
  /home/isim/workspace/linkerd/linkerd2-conformance/tests/ingress/specs.go:18
STEP: Creating ingress-nginx controller
STEP: Injecting linkerd into the ingress controller pods

• Failure [65.663 seconds]
linkerd
/home/isim/workspace/linkerd/linkerd2-conformance/tests/ingress/specs.go:8
  can work with nginx ingress controller [It]
  /home/isim/workspace/linkerd/linkerd2-conformance/tests/ingress/specs.go:18

  could not find a proxy container for deploy/ingress-nginx-controller
  Expected
      <*v1.Container | 0x0>: nil
  not to be nil
...

# went on to run the uninstall test
```

When the script exits, my cluster looks like this:
```
$ k get ns
NAME                 STATUS   AGE
default              Active   31h
ingress-nginx        Active   7m36s  # failed test namespace was left behind, but everything else was deleted
kube-node-lease      Active   31h
kube-public          Active   31h
kube-system          Active   31h
local-path-storage   Active   31h

$ k -n ingress-nginx get po
NAME                                        READY   STATUS      RESTARTS   AGE
ingress-nginx-admission-create-xzhnb        0/1     Completed   0          7m34s
ingress-nginx-admission-patch-wssr4         0/1     Completed   2          7m34s
ingress-nginx-controller-67454c7568-hffz8   2/2     Running     0          6m39s
```

mayankshah1607 replied:
Yep, if you want to be able debug your control plane, you need to have a "universal" control plane installation (set `lifecycle.reinstall` to `false`), and skip the uninstall test (set `lifecycle.uninstall` to false). That way your control plane will stay even if your test fails :smile:

ihcsim replied:
How about a `fail-fast` option where I can decide to skip the rest of the test cases (and don't destroy my control plane), if one fails? I think `go test -failfast` should work?

ihcsim commented on bin/go-test:
```diff
@@ -0,0 +1,10 @@
+    exit
+fi
+
+go test -v ./tests/$1 -ginkgo.v -ginkgo.reportFile=$(pwd)/reports/$2-$1.junit.xml
```
> -v

While the tests are running, the output feels a bit verbose. Every test is repeating the same headers with information on the control plane namespace, cluster domain, CLI path etc. It might be better to just output this information once, at the start of the entire test run. Also, maybe provide a way to enable/disable `-v`.

E.g.,

```
Unspecified linkerd2 control plane namespace - use default value "l5d-conformance" // repeated in every test case
Unspecified cluster domain - using default value "cluster.local" // repeated in every test case
Unspecified path to linkerd2 binary - using default value "/home/isim/.linkerd2/bin/linkerd" // repeated in every test case
linkerd2 binary exists in "/home/isim/.linkerd2/bin/linkerd"- skipping installation
=== RUN   TestLifecycle
Running Suite: Lifecycle
========================
Random Seed: 1594265290  // ?
Will run 1 of 1 specs

 `linkerd install` 
  can install a new control plane
  /home/isim/workspace/linkerd/linkerd2-conformance/tests/lifecycle/specs.go:18
```

ihcsim replied:
What does the `linkerd2 binary exists in "/home/isim/.linkerd2/bin/linkerd"- skipping installation` header that appear before the install test mean? Is the install test being executed after all?

mayankshah1607 replied:
Hmm, the reason why those headers are repeating is because each time a new test suite (`_test.go`) file is run, the config YAML is being re-read and parsed again, as each test suite runs as a separate `go test` process. But I'm sure there's a workaround to just show that output once.

mayankshah1607 replied:
> What does the linkerd2 binary exists in `"/home/isim/.linkerd2/bin/linkerd"- skipping installation` header that appear before the install test mean?

Ah, so that header appears when the tests try to download a linkerd2 binary. Before downloading, it attempts to check if a binary exists in the default `/.linkerd2` directory. If it does, it skips the binary downloading part, which is what "skipping installation" means. I just realised that its misleading, and can easily be confused with "skipping control plane installation". I should change the message to "skipping linkerd2 binary installation".

> Is the install test being executed after all?

Yep, this header actually has nothing to do with the control plane installation, as I mentioned above. The install tests would run regardless. I will change the message :smile:

ihcsim replied:
Using both `go test -v` *and* `-ginkgo.v` as the default verbosity level feels a bit much. IMO, we should minimize the output of passing tests, because they aren't interesting. OTOH, details of failing tests are more important. When my ingress test failed previously, the relevant outputs were buried by those of the passing tests. We can consider omit the `-v` flags by default, and allow users to do `bin/go-test -v` to alter the verbosity level.

ihcsim commented on README.md:
```diff
@@ -1,2 +1,254 @@
+- [`sonobuoy`](https://github.com/mayankshah1607/linkerd2-conformance/tree/master/sonobuoy) contains the items required to be able to run the tests as a [Sonobuoy](https://github.com/vmware-tanzu/sonobuoy) plugin
+- [`utils`](https://github.com/mayankshah1607/linkerd2-conformance/tree/master/utils) contains helper functions that can be used while writing conformance tests
+- [`bin`](https://github.com/mayankshah1607/linkerd2-conformance/blob/master/bin) contains useful helper scripts to build/push the Docker image and running the tests
+- [`test_template`](https://github.com/mayankshah1607/linkerd2-conformance/tree/master/test_template) contains a sample template for adding new tests under the `tests/` folder
```
Does `ginkgo bootstrap` do the same thing?

mayankshah1607 replied:
`ginkgo bootstrap` only provides us with a single file (`<package>_suite_test.go`) which acts as an entrypoint for the ginkgo test suite. Further, `ginkgo generate <testname>` gives us a single test file (`<testname>_test.go`) with a single spec.
The above process is a little naive in the sense that it will run the test cases generated in a random order (because thats how ginkgo deals with multiple `_test.go` files under a single package). We wanted more control over the randomization and better organization over our test cases, which is why we had to have a custom template instead.
[This section](https://github.com/mayankshah1607/linkerd2-conformance#adding-new-tests) explains how our test suites are organized and the significance of the `test_template` directory. :smile: 

Moreoever, each of our `_test.go` files need some logic in the `BeforeSuite` and `AfterSuite` blocks which does not come with `ginkgo generate`. Having a standard test template which satisfies all these requirements seemed a lot more effecient. :smile:

ihcsim replied:
> Moreoever, each of our _test.go files need some logic in the BeforeSuite and AfterSuite blocks 

What are the benefits of separating the tests into different suites? Since these tests need to be executed in a certain order, can we just group them into one test suite? With one test suite,

1. we don't have to introduce an arbitrary test IDs ordering mechanism
1. we don't have to repeat the `BeforeSuite` and `AfterSuite` in every suite
1. the output will be clearer and easier to grok
1. the test template will also be simpler (and probably won't even be necessary)

ihcsim commented on Dockerfile:
```diff
@@ -0,0 +1,23 @@
+COPY go.mod .
+COPY go.sum .
+# Build the test binary
+RUN go test -i ./tests/...
```
Are we just checking for syntax errors here?

mayankshah1607 replied:
The `-i` flag installs the dependencies and compiles our tests for any errors, but does not run them. This was done so that when users are running the tests on Sonouoy, the pod does is only focussed towards running the tests and not install any dependencies.
I forgot to remove `# Build the test binary` comment; it is from when the Dockerfile was a lot different.

ihcsim reviewed (commented):
@mayankshah1607 Good progress! I was able to run the tests locally. See below for some thoughts.

@Pothulapati Feels like this PR is getting really big. Shall we get everything in the branch working and merge it, before adding more new code?

mayankshah1607 commented:
Thanks for the review, @ihcsim ! I've left replies on the comments, and will start working on those fixes. :smile: 

> Shall we get everything in the branch working and merge it, before adding more new code?

Before we do that, I think we need to come to a conclusion regarding Ginkgo and if it is desirable, and whether to move the conformance tests to the upstream linkerd2 repo. This issue is being discussed here - https://github.com/linkerd/linkerd2-conformance/issues/8

Pothulapati commented:
Agree with @mayankshah1607 on sorting out the ginkgo discussion first. But once we do that, and if we think this is the correct way to go (atleast initally), we should merge this PR and raise follow up issues for any concerns.

mayankshah1607 committed:
ingress: fix imports

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
ingress: make IP fetching process retryable

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add .gitignore

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
refactor: make `BeforeEach` skip logic reusable

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
config: Fix YAML field from `linkerdPath` to `linkerdBinaryPath`

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
bin: update scripts to use `#!/usr/bin/env sh`

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
config: nest test case configuration under `testCase` field

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Add workflow for KinD tests

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Remove actions file

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

ihcsim commented on LICENSE:
```diff
@@ -0,0 +1,201 @@
+Apache License
```
The indentation looks a bit off. See the [one in the Linkerd repo](https://github.com/linkerd/linkerd2/blob/main/LICENSE).

ihcsim commented on README.md:
```diff
@@ -1,2 +1,254 @@
+
+_...and a lot more to come_
+
+The Linkerd project is hosted by the Cloud Native Computing Foundation ([CNCF](https://www.cncf.io/)).
```
The more commonly used snippet is:

> Linkerd is a community project and is hosted by the Cloud Native Computing Foundation. Linkerd is committed to open governance.

(See the footer in all the recent blog posts.)

fwiw, I think you can just omit this line since you already hyperlinked it in the first paragraph.

ihcsim commented on README.md:
```diff
@@ -1,2 +1,254 @@
+- [ ] Data plane health checks
+- [ ] Ingress configuration
+
+_...and a lot more to come_
```
I'd present the reader with an opportunity here, with something like:

> If you are interested in helping to extend the test suites, see [Adding new tests](#adding-new-tests) below.

ihcsim commented on README.md:
````diff
@@ -1,2 +1,254 @@
+Run the command below to pull up a sample configuration file, `config.yaml`, and modify it according to your requirements. The tests shall read this YAML file during runtime and run accordingly.
+
+```bash
+$ curl -sL https://raw.githubusercontent.com/mayankshah1607/linkerd2-conformance/master/config.yaml > config.yaml
````
s/mayankshah1607/linkerd

ihcsim commented on README.md:
```diff
@@ -1,2 +1,254 @@
+
+### Using the Sonobuoy CLI
+
+[Sonobuoy](https://github.com/vmware-tanzu/sonobuoy) offers a reliable way to run diagnostic tests in a Kuberenetes cluster. We leverage its [plugin model](https://sonobuoy.io/docs/master/plugins/) to run conformance tests inside a Kubernetes pod.
```
s/Kuberenetes/Kubernetes

ihcsim commented on README.md:
```diff
@@ -1,2 +1,254 @@
+# [Optional] Check the status of the pod
+$ sonobuoy status
+
+# Retrieve the test results results
```
s/test results results/test results

ihcsim commented on README.md:
```diff
@@ -1,2 +1,254 @@
+# Navigate into project directory
+$ cd linkerd2-conformance
+
+# Use the convinence script to run `go run`
```
s/convience/convenient

ihcsim commented on README.md:
```diff
@@ -1,2 +1,254 @@
+# Use the convinence script to run `go run`
+# $ bin/go-test [TEST NAME] [Test ID]
+
+$ bin/go-test inject 01
```
The arbitrary test IDs won't be necessary if we group all the test cases into one test suite, right?

If we really want to keep this, instead of asking the users to manually input an arbitrary test ID, can the script infer it based on the number of tests provided? E.g.,

```
bin/go-test inject # test ID is 01
bin/go-test lifecycle inject ingress # test IDs are 01, 02, 03
```

ihcsim commented on README.md:
```diff
@@ -1,2 +1,254 @@
+
+On completion of the test(s), a `reports/` folder will be created at the root directory of the project, from where the JUnit reports of each individual test can be accessed.
+
+## Adding new tests
```
This is ok for now. But I do want us to think about ways to auto-generate these boiler plate codes as much as possible, in a future PR.

Pothulapati replied:
Also, we should definitely separate this into another doc. but does not have to be in this PR

ihcsim commented on README.md:
```diff
@@ -1,2 +1,254 @@
+| `inject.skip` | Skip proxy injection tests | `false` |
+| `inject.clean` | Delete the resources created for testing proxy injection | `false` |
+| `ingress.skip` | If true, skips all ingress tests | `false` |
+| `ingress.config.controllers` | List of ingress controllers to test. Currently only supports `nginx` | []string |
```
These need to be updated with the `testCase.` prefix.

ihcsim commented on tests/ingress/specs.go:
```diff
@@ -0,0 +1,16 @@
+var _ = ginkgo.Describe("linkerd", func() {
+	_, c := utils.GetHelperAndConfig()
+
+	_ = utils.ShouldTestSkip(c.SkipIngress(), "Skipping ingress tests")
```
Does this happen before or after `BeforeSuite`, `TestIngress` etc.? I'd think we want this check to happen sooner before setting up emojivoto etc.

ihcsim commented on tests/ingress/tests.go:
```diff
@@ -0,0 +1,104 @@
+	ginkgo.By("Creating ingress-nginx controller")
+	_, err := h.Kubectl("", "apply", "-f", "testdata/controllers/nginx.yaml")
+
+	gomega.Expect(err).Should(gomega.BeNil(), fmt.Sprintf("failed to create controller: %s", utils.Err(err)))
```
I don't think you need the `utils.Err(err)` at the end; should be able to just do `fmt.Sprintf("failed to create controller: %s", err)`. Otherwise, it feels simpler to just do the good old:
```
if err != nil {
  return err
}
```

ihcsim reviewed (commented):
Feels like this is very close to getting merged. My main thought here is whether it's necessary to separate the tests into separate test suites, since these tests need to be executed in a certain order. The multiple test suites seem to complicate things, resulting in more boiler plate code, arbitrary test IDs ordering mechanism, a code template that future contributors need to follow etc. Can we simplify them by grouping all the tests in one test suite?

mayankshah1607 commented:
@ihcsim The points you listed make absolute sense. In fact, we initially started out as a single test suite. The main problem we faced was the output which Sonobuoy gave us (via `sonobuoy results --mode dump`) was very unorganized. For example, all the test cases were nested under a single parent block:
```yaml
name: linkerd2-conformance
status: passed
meta:
  type: summary
items:
- name: report.xml
  status: passed
  meta:
    file: results/global/report.xml
    type: file
  items:
  - name: Linkerd2 Conformance Validation
    status: passed
    items:
    - name: 'lifecycle:`linkerd install` can install a new control plane'
      status: passed
    - name: '  `lifecycle: linkerd upgrade` can upgrade CLI'
      status: skipped
    - name: '  `lifecycle: linkerd upgrade` can upgrade control-plane'
      status: skipped
    - name: '  `inject: linkerd inject` can perform manual injection'
      status: passed
    - name: '  `inject: linkerd inject` can inject proxy container into pods'
      status: passed
    - name: '  `inject: linkerd inject` can override pod level proxy config with namespace
        level config'
      status: passed
    - name: '  `inject: linkerd inject` should delete all resources created during testing'
      status: passed
    - name: '  `inject: linkerd install` can uninstall control plane'
      status: passed
```

When separating the test cases into individual test suites (`_test.go` files), multiple JUnit reports were produced, one for each test suite (under the `reports/` directory). Sonobuoy has the ability to aggregate multiple JUnit reports, giving each report a separate section under the Sonobuoy results YAML, making it look a lot more organized. For example, here is what the output of `sonobuoy results` if we have a test suite for each feature:
```yaml
name: linkerd2-conformance
status: passed
meta:
  type: summary
items:
- name: report.xml
  status: passed
  meta:
    file: results/global/report.xml
    type: file
  items:
  - name: Life-Cycle
    status: passed
    items:
    - name: '  `linkerd install` can install a new control plane'
      status: passed
    - name: '  `linkerd upgrade` can upgrade CLI'
      status: skipped
    - name: '  `linkerd upgrade` can upgrade control-plane'
      status: skipped
  - name: Inject
    status: passed
    items:
    - name: '  `linkerd inject` can perform manual injection'
      status: passed
    - name: '  `linkerd inject` can inject proxy container into pods'
      status: passed
    - name: '  `linkerd inject` can override pod level proxy config with namespace
        level config'
      status: passed
    - name: '  `linkerd inject` should delete all resources created during testing'
      status: passed
    - name: '  `linkerd install` can uninstall control plane'
      status: passed
```
Considering that Sonobuoy would be our primary way of running these tests, @Pothulapati and I decided to give the Sonobuoy results part some extra weightage. 

Another disadvantage that I can think of with having a single test suite, is that we would then no longer have the ability to perform pre-testing setup such as installing the control plane or a sample application. We could very well make these a part of the test case itself (i.e inside the `Describe` block, as opposed to `BeforeSuite`), but then our JUnit report is going to be pulluted with the extra verbosity this will bring; i.e., frequently seeing messages like `"<testname>: installing sample application emojivoto"`, or `"<testname>: installing control plane"`, etc.

Please let me know what you think. Thanks again. :smile:

Pothulapati commented:
@ihcsim The boiler plate bash definitely felt like a problem, but we could not find a way to get the nested outputs without separating tests :\

mayankshah1607 committed:
major refactor: wrap all tests into a single test suite

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
update Dockerfile and Sonobuoy items

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
ingress: Add emojivoto test spec

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
ingress: increase emojivoto request timeout

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
ingress: make pinging retryable

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

ihcsim commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+# install kubectl
+RUN curl -LO https://storage.googleapis.com/kubernetes-release/release/`curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt`/bin/linux/amd64/kubectl
+RUN chmod +x ./kubectl
+RUN mv ./kubectl /usr/local/bin
```
You can `&&` these to avoid the extra layers.

ihcsim commented on conformance_test.go:
```diff
@@ -0,0 +1,33 @@
+func TestMain(m *testing.M) {
+	h, c := utils.GetHelperAndConfig()
+
+	// install linkerd binary
```
Should this be moved to line 21?

mayankshah1607 committed:
sonobuoy: update scipts

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: update README

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
update license

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update README.md

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: fix anchor links

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

ihcsim commented on README.md:
```diff
@@ -1,2 +1,239 @@
+
+## Table of Contents
+
+- [Repository Structure](https://github.com/mayankshah1607/linkerd2-conformance#repository-structure)
```
s/mayankshah1607/linkerd

ihcsim commented on README.md:
```diff
@@ -1,2 +1,239 @@
+
+[Sonobuoy](https://github.com/vmware-tanzu/sonobuoy) offers a reliable way to run diagnostic tests in a Kubernetes cluster. We leverage its [plugin model](https://sonobuoy.io/docs/master/plugins/) to run conformance tests inside a Kubernetes pod.
+
+
```
Extra line break.

ihcsim commented on README.md:
```diff
@@ -1,2 +1,239 @@
+
+#### 2. Writing the tests
+   
+`spec.go` must contain a single (if required, more) exported function that returns a `ginkgo.Describe` block that holds the specs. This function must be named `Runl5dFeatureTests`. 
```
>  must contain a single (if required, more) 

As in, _at least one_?

ihcsim commented on README.md:
````diff
@@ -1,2 +1,239 @@
+```go
+func runPrimaryTests() bool {
+	return ginkgo.Describe("", func() {
+    // ...test initialisation logic here
+
+		// Bring main tests into scope
+    _ = inject.RunInjectTests()
+    _ = l5dFeature.Runl5dFeatureTests() // call your test here
+
+		// ...post testing logic here
````
Mind fixing up the identation here? Thanks.

ihcsim reviewed (approved):
This LGTM! Thanks for simplifying it. Just some comments on the README, then we should be able to merge this :+1:.

ihcsim commented:
Also, please ensure all lines in the README is wrapped to 80 characters.

mayankshah1607 committed:
Fix linting

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Fix markdown linting

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
Update Dockerfile

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: Fix indentation

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

mayankshah1607 committed:
docs: s/mayankshah1607/linkerd

Signed-off-by: Mayank Shah <mayankshah1614@gmail.com>

Pothulapati commented on README.md:
````diff
@@ -1,2 +1,315 @@
+curl -sL $CONFIG > config.yaml
+```
+
+### Configuration options
````
I would keep the configuration options later in the file, as users don't have to touch this in the default case.

Pothulapati reviewed (approved):
This LGTM, and most of my concerns were regarding docs. So, Let's create another issue something like `docs restructure` and follow up there?
